src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker_IVsSolutionLoadEvents.cs

```diff
@@ -51,6 +95,217 @@ int IVsSolutionLoadEvents.OnAfterBackgroundSolutionLoadComplete()
+        private IWorkspaceProjectContext CreateProjectFromArgumentsAndReferences(
+            IWorkspaceProjectContextFactory workspaceProjectContextFactory,
+            string projectFilename,
+            IReadOnlyDictionary<string, DeferredProjectInformation> allProjectInfos,
+            IReadOnlyDictionary<string, string> targetPathsToProjectPaths)
+        {
+            var languageName = GetLanguageOfProject(projectFilename);
+            if (languageName == null)
+            {
+                return null;
+            }
+
+            DeferredProjectInformation projectInfo;
+            if (!allProjectInfos.TryGetValue(projectFilename, out projectInfo))
+            {
+                return null;
+            }
+
+            var commandLineParser = _workspaceServices.GetLanguageServices(languageName).GetService<ICommandLineParserService>();
+            var projectDirectory = Path.GetDirectoryName(projectFilename);
+            var commandLineArguments = commandLineParser.Parse(
+                projectInfo.CommandLineArguments,
+                projectDirectory,
+                isInteractive: false,
+                sdkDirectory: RuntimeEnvironment.GetRuntimeDirectory());
+
+            // TODO: Should come from sln file?
+            var projectName = Path.GetFileNameWithoutExtension(projectFilename);
+            var projectId = GetOrCreateProjectIdForPath(projectFilename, projectName);
+
+            AbstractProject project;
+            if (_projectMap.TryGetValue(projectId, out project))
+            {
+                return project as IWorkspaceProjectContext;
+            }
+
+            OutputToOutputWindow($"\tCreating '{projectName}':\t{commandLineArguments.SourceFiles.Length} source files,\t{commandLineArguments.MetadataReferences.Length} references.");
+            var solution5 = _serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution5;
+            var projectGuid = solution5.GetGuidOfProjectFile(projectFilename);
+            var projectContext = workspaceProjectContextFactory.CreateProjectContext(
+                languageName,
+                projectName,
+                projectFilename,
+                projectGuid: projectGuid,
+                hierarchy: null,
+                binOutputPath: projectInfo.TargetPath);
+
+            projectContext.SetOptions(projectInfo.CommandLineArguments.Join(" "));
+
+            foreach (var sourceFile in commandLineArguments.SourceFiles)
+            {
+                projectContext.AddSourceFile(sourceFile.Path);
+            }
+
+            foreach (var sourceFile in commandLineArguments.AdditionalFiles)
+            {
+                projectContext.AddAdditionalFile(sourceFile.Path);
+            }
+
+            var addedProjectReferences = new HashSet<string>();
+            foreach (var projectReference in projectInfo.ProjectReferences)
+            {
+                var projectReferencePath = projectReference.Substring("/ProjectReference:".Length);
```

Comment:
From the AnyCode binaries.  Moved into the `DeferredProjectWorkspaceService` so it's hidden away from consumers.